### PR TITLE
MAINT: Change c*np.ones(...) to np.full(..., c, ...) in many places

### PR DESCRIPTION
**Pull request by endolith**
#### What does this implement/fix?
When c is a scalar, c * np.ones(...) and np.full(..., c, ...) produce the same output, but np.full is [faster](https://stackoverflow.com/a/45006691/125507)
and clearer in intent, so I (somewhat mechanically) replaced a lot of instances.

```
timeit np.ones((100, 1000)) / 12.
725 µs ± 37.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

timeit np.full((100, 1000), 1 / 12.)
75.2 µs ± 7.44 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

I did not change ones inside benchmarks or where I wasn't sure that it's always a scalar.

### CODE REVIEW

**Comment by larsoner** [comment]
Looks like a good idea to me, let me know when everything is green and I should look @endolith

**Comment by endolith** [comment]
@larsoner It's green now.

**Review by larsoner — APPROVED** [review]
Looks good to me, +1 for merge

**Comment by peterbell10** [comment]
>When c is a scalar, c * np.ones(...) and np.full(..., c, ...) produce the same output

There is at least one difference worth mentioning; the `dtype` may be different. e.g.
```python
>>> (5 * np.ones(10)).dtype
dtype('float64')
>>> np.full(10, 5).dtype
dtype('int64')
```

File: scipy/stats/tests/test_stats.py [hunk 1]
```diff
@@ -2110,7 +2110,7 @@ def test_nanpolicy(self):
                 _check_warnings(w, RuntimeWarning, 3)
             else:
                 assert_equal(stats.iqr(x, nan_policy='omit'), 7.5)
-                assert_equal(stats.iqr(x, axis=0, nan_policy='omit'), 5 * np.ones(5))
+                assert_equal(stats.iqr(x, axis=0, nan_policy='omit'), np.full(5, 5))
```
Comment (endolith) [inline scipy/stats/tests/test_stats.py]: assert_equal doesn't care about type

File: scipy/stats/tests/test_stats.py [hunk 2]
```diff
@@ -3476,7 +3476,7 @@ def test_describe_scalar(self):
         assert_array_almost_equal(kurt, -3.0, decimal=13)
 
     def test_describe_numbers(self):
-        x = np.vstack((np.ones((3,4)), 2 * np.ones((2,4))))
+        x = np.vstack((np.ones((3,4)), np.full((2, 4), 2)))
```
Comment (endolith) [inline scipy/stats/tests/test_stats.py]: Combined with float array so doesn't matter

**Review by endolith — COMMENTED** [review]
They all seem fine as far as type changes. Anything else to look for?

**Comment by larsoner** [comment]
I tried to be conscious of the types while looking, but one more set of reviewer eyes would be good

File: scipy/fft/tests/test_numpy.py [hunk 3]
```diff
@@ -279,19 +279,19 @@ def worker(args, q):
                 'Function returned wrong value in multithreaded context')
 
     def test_fft(self):
-        a = np.ones(self.input_shape) * 1+0j
+        a = np.full(self.input_shape, 1+0j)
```
Comment (peterbell10) [inline scipy/fft/tests/test_numpy.py]: Nit: these are all still filling with ones.
```suggestion
        a = np.ones(self.input_shape, dtype=np.complex128)
```

File: scipy/optimize/nonlin.py [hunk 4]
```diff
@@ -1227,7 +1227,7 @@ def __init__(self, alpha=None):
 
     def setup(self, x, F, func):
         GenericBroyden.setup(self, x, F, func)
-        self.d = np.ones((self.shape[0],), dtype=self.dtype) / self.alpha
+        self.d = np.full((self.shape[0],), 1 / self.alpha, dtype=self.dtype)
```
Comment (peterbell10) [inline scipy/optimize/nonlin.py]: Might be different if `self.dtype` can be integral.

File: scipy/stats/tests/test_discrete_basic.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ def test_rvs_broadcast(dist, shape_args):
     for k in range(nargs):
         shp = (k + 3,) + (1,)*(k + 1)
         param_val = shape_args[k]
-        allargs.append(param_val*np.ones(shp, dtype=np.array(param_val).dtype))
+        allargs.append(np.full(shp, param_val, dtype=np.array(param_val).dtype))
```
Comment (peterbell10) [inline scipy/stats/tests/test_discrete_basic.py]: Looks like this code explicity wants the same dtype as `param_val`.
```suggestion
        allargs.append(np.full(shp, param_val))
```

**Review by peterbell10 — APPROVED** [review]
I had a look through and I agree that where the dtype has changed, it still looks okay. A few nitpicks but otherwise looks good.

File: scipy/optimize/nonlin.py (hunk 4 above)
Comment (larsoner) [reply scipy/optimize/nonlin.py]: In this case I suspect that it's a bug and they meant for `self.d` to have the same dtype, even though dividing by `self.alpha` might change it. So I'm okay with it as is.

**Comment by larsoner** [comment]
Thanks @endolith

**Pull request merged**